test/parallel/test-http-reuse-socket.js

```diff
@@ -0,0 +1,73 @@
+// Copyright Joyent, Inc. and other Node contributors.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to permit
+// persons to whom the Software is furnished to do so, subject to the
+// following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
+// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+// USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+'use strict';
+const common = require('../common');
+const http = require('http');
+const assert = require('assert');
+const Countdown = require('../common/countdown');
+
+// The HEAD:204, GET:200 was the most pathological test case I could find.
+// GETs following a 204 response with a content-encoding header failed.
+// Responses without bodies and without content-length or encoding caused
+//   the socket to be closed.
+const codes = [204, 200, 200];
+const methods = ['HEAD', 'HEAD', 'GET'];
+
+const sockets = [];
+const agent = new http.Agent();
+agent.maxSockets = 1;
+
+const countdown = new Countdown(codes.length, () => server.close());
+
+const server = http.createServer(common.mustCall((req, res) => {
+  const code = codes.shift();
+  assert.strictEqual(typeof code, 'number');
+  assert.ok(code > 0);
+  res.writeHead(code, {});
+  res.end();
+}, codes.length));
+
+function nextRequest() {
+
+  const request = http.request({
+    port: server.address().port,
+    path: '/',
+    agent: agent,
+    method: methods.shift()
+  }, common.mustCall((response) => {
+    response.on('end', common.mustCall(() => {
+      if (countdown.dec()) {
+        nextRequest();
+      }
+      assert.strictEqual(sockets.length, 1);
```

Comment:
For my edification: isn't this tautologically true because of the `agent.maxSockets = 1` on line 37?

Reply:
It's my understanding that `agent.maxSockets` refers to the total concurrent sockets, my aim with this test is to record every socket opened and count them.

Running (a corrected version that I'll commit shortly) on master yields:
`=== release test-http-reuse-socket2 ===
Path: parallel/test-http-reuse-socket2
assert.js:103
  throw new AssertionError(obj);
  ^

AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:

2 !== 1

    at IncomingMessage.<anonymous> (/home/joseph/Source/node/test/parallel/test-http-reuse-socket2.js:39:14)
    at IncomingMessage.<anonymous> (/home/joseph/Source/node/test/common/index.js:365:15)
    at IncomingMessage.emit (events.js:326:22)
    at endReadableNT (_stream_readable.js:1261:12)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  generatedMessage: true,
  code: 'ERR_ASSERTION',
  actual: 2,
  expected: 1,
  operator: 'strictEqual'
}`